cmd/solvercli/cmd/balances.go

```diff
@@ -0,0 +1,245 @@
+}
+
+type DenomDetail struct {
+	Amount          string `json:"amount"`
+	Decimals        uint8  `json:"decimals"`
+	FormattedAmount string `json:"formatted_amount"`
+	Price           string `json:"price"`
+	ValueUSD        string `json:"value_usd"`
+}
+
+type ChainBalance struct {
+	ChainID    string
+	AssetDenom string
+	Balance    *big.Int
+	Symbol     string
+	Decimals   uint8
+	PriceUSD   *big.Float
+	ValueUSD   *big.Float
+}
+
+type ChainGasBalance struct {
+	ChainID           string
+	Balance           *big.Int
+	Symbol            string
+	Decimals          uint8
+	WarningThreshold  *big.Int
+	CriticalThreshold *big.Int
+}
+
+var balancesCmd = &cobra.Command{
+	Use:   "balances",
+	Short: "Show current on-chain balances (USDC and gas tokens)",
+	Long: `Show current on-chain balances for USDC and gas tokens across all configured chains.
+    
+Example:
+    ./build/solvercli balances \
+    --custom-assets '{"osmosis-1":["uosmo","uion"],"celestia-1":["utia"]}'`,
+	Run: func(cmd *cobra.Command, args []string) {
+		ctx := setupContext(cmd)
+		usdcBalances, gasBalances, customBalances, totalUSDCBalance, totalCustomAssetsUSDValue, err := getBalances(ctx, cmd)
+		if err != nil {
+			lmt.Logger(ctx).Fatal("Failed to get balances", zap.Error(err))
+		}
+
+		fmt.Println("\nOn-Chain Balances:")
+		fmt.Println("------------------")
+
+		for chainID, usdc := range usdcBalances {
+			gas := gasBalances[chainID]
+			fmt.Printf("\nChain: %s\n", chainID)
+			fmt.Printf("  USDC Balance: %s USDC\n", normalizeBalance(usdc.Balance, CCTP_TOKEN_DECIMALS))
+			fmt.Printf("  Gas Balance: %s %s\n", normalizeBalance(gas.Balance, gas.Decimals), gas.Symbol)
+
+			if gas.Balance.Cmp(gas.CriticalThreshold) < 0 {
+				fmt.Printf("  ⚠️  Gas balance below critical threshold!\n")
+			} else if gas.Balance.Cmp(gas.WarningThreshold) < 0 {
+				fmt.Printf("  ⚠️  Gas balance below warning threshold\n")
+			}
+
+			// Print custom assets if available
+			if assets, ok := customBalances[chainID]; ok {
+				for _, asset := range assets {
+					fmt.Printf("  %s Balance: %s %s (%.2f USD)\n",
+						asset.AssetDenom,
+						normalizeBalance(asset.Balance, asset.Decimals),
+						asset.AssetDenom,
+						asset.ValueUSD)
+				}
+			}
+		}
+
+		fmt.Printf("\nTotals Across All Chains:")
+		fmt.Printf("\n------------------------\n")
+		fmt.Printf("Total USDC Balance: %s USDC\n", normalizeBalance(totalUSDCBalance, CCTP_TOKEN_DECIMALS))
+		if totalCustomAssetsUSDValue.Cmp(big.NewFloat(0)) > 0 {
+			fmt.Printf("Total Custom Assets USD Value: %.2f USD\n", totalCustomAssetsUSDValue)
+		}
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(balancesCmd)
+	balancesCmd.Flags().String("custom-assets", "", "JSON map of chain IDs to denom arrays, e.g. '{\"osmosis\":[\"uosmo\",\"uion\"],\"celestia\":[\"utia\"]}'")
+}
+
+func getBalances(ctx context.Context, cmd *cobra.Command) (map[string]*ChainBalance, map[string]*ChainGasBalance, map[string][]*ChainBalance, *big.Int, *big.Float, error) {
```

Comment:
this feels like it should be a couple different functions, from looking at it, maybe a function to gather up what chain + asset pairs the user wants to get the balances of, one to get the actual balances response, and one to parse? maybe some other split makes more sense, but usually when there are 3 or more returns the func should probably be broken up

Reply:
i refactored the logic a bit here, take a look and lmk if its easier to follow now :D